### DISCUSSION TITLE
Proposal: Factory initializer syntax

### DISCUSSION THREAD

**Discussion by ReneeGA2020** [dotnet/csharplang#8331] [General] [answered]
Sometimes we use a factory to return a descendant instance of a certain type, but when you have a bunch of required properties, the parameter list of the factory function would become insanely long, which just negates the benefit of C#'s initializer syntax.

```
   public Item FactoryCreateItemt(Scene scene, Player player , vector point, string name....)
   {
        .......
        case A:
        return new Item()
       {
          Scene = scene,
          Player = player,
          Point = point,
          Name = name
          ....
       }
       case B:
       return new ItemB()
       {
          Scene = scene,
          Player = player,
          Point = point,
          Name = name
          ....
       }
   }
```

The idea would be like, put a 'new' keyword after the return type of the factory function, which indicates that you will have to return with a new constructed object of the returned type.

```
   public Item new FactoryCreateItemt()
   {
          case A:
          // Basically equivalent to 'return new Item()', but it will append the initializer of the caller code to the object.
          new return Item();
         
         case B:
         // Basically equivalent to 'return new ItemB()', but it will append the initializer of the caller code to the object.
         new return ItemB();
   }
  ...
  var item =  new FactoryCreateItemt(){
          Scene = scene,
          Player = player,
          Point = point,
          Name = name
          ....
   };
```
So that we can treat the factory function as if it’s a real constructor and can use initializer syntax after it. By using 'TReturn new' instead of 'new TReturn', it won’t conflict with the existing ‘hide inherited function’ syntax.

The idea would be that you would have to do ‘new return …’ instead of 'return new ...' in the factory (similar to the syntax of ‘yield return’), as long as you put a ‘new’ keyword after the return type in the definition.

The factory function would have to allow you to do ‘return new Item()’ even if you didn’t specify the required fields in the return statement. In that case, the responsibility of setting up the required properties would fall to the caller code of the factory.

Perhaps you can still do 'new return Item(){ ... }' initializer in the new return line, and the caller's initializer should add up to it. And naturally, you would also be able to use attributes like [SetsRequiredMembers]:

```
[SetsRequiredMembers]
public Item new FactoryCreateItem()
{
    // ...
}
```

So that the caller wouldn’t have to set the required fields, but would still have the option to use initializer syntax.

**Comment by ufcpp** [comment, marked as answer]
See Factories / Factory methods sections in https://github.com/dotnet/csharplang/discussions/3707 https://github.com/dotnet/csharplang/blob/cf67a0f8ca0ab3b8da79a2f9ea8cb23f253b4e5a/meetings/2020/LDM-2020-04-20.md https://github.com/dotnet/csharplang/blob/cf67a0f8ca0ab3b8da79a2f9ea8cb23f253b4e5a/meetings/2020/LDM-2020-05-04.md
